[@property] Support list values

anttijk opened the pull request:
#### f266283a31249f27645c470313b8ee346c4f0daa
<pre>
[@property] Support list values
<a href="https://bugs.webkit.org/show_bug.cgi?id=249216">https://bugs.webkit.org/show_bug.cgi?id=249216</a>
&lt;rdar://problem/103300866&gt;

Reviewed by Chris Dumez.

Consume values with &lt;type&gt;+ and &lt;type&gt;# syntax definitions.

* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/register-property-syntax-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/registered-property-computation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/registered-property-initial-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/typedom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/var-reference-registered-properties-expected.txt:
* Source/WebCore/css/CSSCustomPropertyValue.cpp:
(WebCore::CSSCustomPropertyValue::equals const):
(WebCore::CSSCustomPropertyValue::customCSSText const):
* Source/WebCore/css/CSSCustomPropertyValue.h:

Add list value support.
Handle ident values with SyntaxValue too.
Use URL type for URLs.

* Source/WebCore/css/CSSValue.cpp:
(WebCore::CSSValue::collectDirectComputationalDependencies const):
(WebCore::CSSValue::collectDirectRootComputationalDependencies const):

Collect from lists too.

* Source/WebCore/css/DOMCSSRegisterCustomProperty.cpp:
(WebCore::DOMCSSRegisterCustomProperty::registerProperty):

Test for null initial value string instead of empty (empty string needs to be considered as an initial value).
Remove unnecessary collectDirect*ComputationalDependencies calls.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::consumeCustomPropertyValueWithSyntax):

Consume lists.

(WebCore::CSSPropertyParser::collectParsedCustomPropertyValueDependencies):
(WebCore::CSSPropertyParser::parseTypedCustomPropertyValue):

Resolve syntax values in lists.

Canonical link: <a href="https://commits.webkit.org/257844@main">https://commits.webkit.org/257844@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1c938f30b8f35c9c678f0917cca765ae4f505368

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100192 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9359 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33266 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109519 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169752 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104184 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10257 "Built successfully") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92613 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107408 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105961 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/7757 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91032 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/34444 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89668 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/22427 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/77386 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3118 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/23943 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3096 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9224 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43431 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5383 "'git push ...'") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4928 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/css/CSSCustomPropertyValue.cpp:
```diff
@@ -89,6 +93,15 @@ String CSSCustomPropertyValue::customCSSText() const
+            StringBuilder builder;
+            auto separator = separatorCSSText(syntaxValueList.separator);
+            for (auto& syntaxValue : syntaxValueList.values) {
+                if (&syntaxValue != &syntaxValueList.values.first())
```
I personally think this would have looked a bit nicer:
`if (!builder.isEmpty())`

Pretty common pattern in WebKit code base too.

cdumez commented on Source/WebCore/css/DOMCSSRegisterCustomProperty.cpp:
```diff
@@ -60,7 +60,7 @@ ExceptionOr<void> DOMCSSRegisterCustomProperty::registerProperty(Document& docum
 
         HashSet<CSSPropertyID> dependencies;
-        CSSPropertyParser::collectParsedCustomPropertyValueDependencies(*syntax, false, dependencies, tokenizer.tokenRange(), strictCSSParserContext());
+        CSSPropertyParser::collectParsedCustomPropertyValueDependencies(*syntax, true, dependencies, tokenizer.tokenRange(), strictCSSParserContext());
```
Would be nice to put in an inline comment `/* */` to indicate what this boolean means and switch to an enum class for better readability.

cdumez reviewed (approved):
LGTM

anttijk committed:
[@property] Support list values
https://bugs.webkit.org/show_bug.cgi?id=249216
<rdar://problem/103300866>

Reviewed by Chris Dumez.

Consume values with <type>+ and <type># syntax definitions.

* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/register-property-syntax-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/registered-property-computation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/registered-property-initial-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/typedom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/var-reference-registered-properties-expected.txt:
* Source/WebCore/css/CSSCustomPropertyValue.cpp:
(WebCore::CSSCustomPropertyValue::equals const):
(WebCore::CSSCustomPropertyValue::customCSSText const):
* Source/WebCore/css/CSSCustomPropertyValue.h:

Add list value support.
Handle ident values with SyntaxValue too.
Use URL type for URLs.

* Source/WebCore/css/CSSValue.cpp:
(WebCore::CSSValue::collectDirectComputationalDependencies const):
(WebCore::CSSValue::collectDirectRootComputationalDependencies const):

Collect from lists too.

* Source/WebCore/css/DOMCSSRegisterCustomProperty.cpp:
(WebCore::DOMCSSRegisterCustomProperty::registerProperty):

Test for null initial value string instead of empty (empty string needs to be considered as an initial value).
Remove unnecessary collectDirect*ComputationalDependencies calls.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::consumeCustomPropertyValueWithSyntax):

Consume lists.

(WebCore::CSSPropertyParser::collectParsedCustomPropertyValueDependencies):
(WebCore::CSSPropertyParser::parseTypedCustomPropertyValue):

Resolve syntax values in lists.

Canonical link: https://commits.webkit.org/257844@main
